subsys/bootloader/image/CMakeLists.txt

```diff
@@ -4,59 +4,53 @@
 # SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
 #
 
-zephyr_include_directories(${CMAKE_CURRENT_LIST_DIR}/../include)
-if (NOT DEFINED CONFIG_IS_SECURE_BOOTLOADER)
-  add_subdirectory(bl_crypto_client)
+# This CMakeLists.txt is executed by the root application only when
+# B0 (SECURE_BOOT) is enabled. First, figure out what image will be
+# booted by B0, and set the required properties for that image to be
+# bootable by B0.
+
+if (CONFIG_BOOTLOADER_MCUBOOT)
+  set(image_to_boot mcuboot_)
+  # Partition S1 is only enabled when MCUBoot is enabled.
+  set(s1_addr $<TARGET_PROPERTY:partition_manager,PM_S1_ADDRESS>)
+elseif(CONFIG_SPM)
+    set(image_to_boot spm_)
 endif()
-set(NRF_BOOTLOADER_SCRIPTS ${NRF_DIR}/scripts/bootloader)
-set(PROVISION_HEX_NAME     provision_data.hex)
-set(PROVISION_HEX          ${PROJECT_BINARY_DIR}/${PROVISION_HEX_NAME})
-set(SIGNED_KERNEL_HEX      ${PROJECT_BINARY_DIR}/signed_by_b0_${KERNEL_HEX_NAME})
 
-include(${CMAKE_CURRENT_LIST_DIR}/../cmake/debug_keys.cmake)
-include(${CMAKE_CURRENT_LIST_DIR}/../cmake/sign.cmake)
+# Include a kconfig file which enables CONFIG_FW_METADATA in the image
+# which is booted by B0.
+set(old_conf ${${image_to_boot}OVERLAY_CONFIG})
+set(${image_to_boot}OVERLAY_CONFIG
+  "${old_conf} ${CMAKE_CURRENT_SOURCE_DIR}/fw_metadata.conf"
+  CACHE STRING "" FORCE)
 
-if(DEFINED DT_FLASH_AREA_S0_OFFSET)
-  set(S0_ADDR ${DT_FLASH_AREA_S0_OFFSET})
-else()
-  set(S0_ADDR 0x8000)
+# Partition S0 will always contain the image to be booted by B0.
+set(s0_addr $<TARGET_PROPERTY:partition_manager,PM_S0_ADDRESS>)
+if (s1_addr)
+  set(s1_arg --s1-addr ${s1_addr})
 endif()
 
-if(DEFINED DT_FLASH_AREA_S1_OFFSET)
-  set(S1_ADDR ${DT_FLASH_AREA_S1_OFFSET})
-else()
-  set(S1_ADDR 0xc000)
-endif()
+# Include crypto client to allow booted image to re-use B0's crypto.
+zephyr_include_directories(${CMAKE_CURRENT_LIST_DIR}/../include)
+add_subdirectory(bl_crypto_client)
```

Comment:
Which images get this subdirectory added?

Reply:
The root app, which might not be correct. It should actually be possible for all non-b0 images to include this (conditional on some kconfig). 

I will see if I can come up with a better solution.

Reply:
Check now, added separate kconfig which allows any image to enable the SECURE_BOOT_CRYPTO_CLIENT.